[2단계 - 페이먼츠] 오거스(오명석) 미션 제출합니다.

ohgus opened the pull request:
안녕하세요 시지프! 주말 잘보내셨나요? 😁
2단계도 잘 부탁드립니다! 🙇
이번 2단계를 진행하면서 개인적으로 과하게 욕심을 부리다 시간이 모자르는 상황이 되었습니다. 그래서 원하는 수준의 리팩토링을 진행하지 못했고 최종적으로 시간안에 구현을 완료하는 것을 목표로 마무리하게 되었습니다.
그리고 최종적으로 시간이 좀 모자라서 pr에 작성하고 싶은 내용들을 정리하지 못했습니다... 혹시 첫 리뷰 이후에 궁금한 부분들을 더 정리해서 질문해도 괜찮을까요? 

## 🎯 페이먼츠

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계
- 재사용 가능한 Input Component를 개발한다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 테스트한다.
- 카드 정보를 효과적으로 렌더링 하기 위한 상태 관리를 경험한다.

### 2단계
- 다양한 Form 구성 요소들간의 상태를 효율적으로 관리한다.
- hooks API를 이용하여 상태 관리 로직을 구현하한다.
- custom hooks를 생성하여 Form 관리 로직을 컴포넌트에서 분리하고 재사용한다.
- Controlled & Uncontrolled Components에 입각하여 Form을 핸들링한다.

---

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항(코드 컨벤션, 에러 핸들링 등)을 준수했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - [payments](https://ohgus.github.io/react-payments/)
  - [storybook](https://67ff92a6eff4aa02ca30799d-ewfnfdubof.chromatic.com/?path=/docs/foundations-colors--docs)

### 리뷰 요청 & 논의하고 싶은 내용

#### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

### Form tag 사용

이번 미션에서는 실제로 서버로 데이터를 제출하는 기능이 포함되어 있지 않았기 때문에, 초반에는 별도의 form 태그를 사용하지 않고 각 input 요소들을 독립적으로 관리했습니다.

하지만 기능의 흐름을 다시 정리해보니,
**“모든 입력 데이터를 작성 완료한 후 하나의 버튼 클릭으로 카드 정보를 등록한다”** 는 시나리오가 명확히 드러났습니다.

이를 고려했을 때,
	•	단순 구현 관점에서는 form이 필수적이지 않지만,
	•	행동(Submit) 단위로 입력 그룹을 묶는 의미,
	•	웹 접근성(semantics) 측면에서 논리적 구조 제공,
	•	키보드 사용성(Enter 키 입력 시 자연스럽게 submit 처리 가능성)

등을 모두 고려하면 form 태그를 사용하는 것이 더 시멘틱하고,
사용성과 확장성 측면에서도 자연스러운 선택이라고 판단했습니다.

#### 2) 이번 리뷰를 통해 논의하고 싶은 부분

### hook 분리 & 재사용

현재 작업 중, Hook의 분리와 재사용성에 대해 깊게 고민하게 되었습니다.

완전히 다른 데이터 구조라면 당연히 별도의 Hook으로 분리하는 것이 자연스럽다고 생각합니다. 하지만 데이터 성격이 유사하거나, 약간의 구조 타협만으로 하나의 Hook 안에서 포괄할 수 있을 경우, 다음과 같은 두 갈래 사이에서 고민하고 있습니다.
	•	1. 재사용성과 일관성을 고려해 하나의 Hook으로 통합하여 관리하는 편이 나은지
	•	2. 데이터의 본질적 의미와 역할을 기준으로 명확히 분리하여 각각의 Hook을 유지하는 것이 좋은지

특히, “구조 타협”을 감수하면서까지 하나로 묶는 것이 장기적인 관점에서도 올바른 선택일지, 아니면 초기에 조금 더 분리 비용을 감수하더라도 데이터 성격에 맞게 관리하는 것이 더 바람직한지에 대한 고민이 깊어졌습니다.

이런 상황에서 시지프라면 어떤 기준을 가장 우선순위에 둘지 궁금합니다.

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도만이 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. Form 상태 관리 & Custom Hook 분리

- 반복되는 로직을 custom hook으로 분리했는가?
- hook 내부와 UI 컴포넌트의 **역할이 명확하게 분리**되어 있는가?
- 상태 흐름이 직관적이며, Form 전체를 일관되게 관리할 수 있는 구조인가?

### 2. 입력 UI 흐름과 UX

- 카드 번호 입력 시 필드 간 **자동 포커싱 이동**이 자연스럽게 동작하는가?
- 숫자만 입력 가능한 필드에서 제한, 에러 메시지, 유효성 피드백 등 **사용자 경험이 충분히 고려**되었는가?

### 3. 컴포넌트 구조 및 재사용성

- 컴포넌트가 명확한 **역할과 책임**을 가지며 과도하게 분리되거나 중첩되어 있지 않은가?
- Input, Button 등 재사용 가능한 컴포넌트가 잘 정의되어 있는가?

### 4. 상태 기반 유효성 검사 및 확인 버튼 활성화

- 모든 필드가 유효할 때만 확인 버튼이 **정확히 활성화/비활성화**되는가?
- 유효성 검사의 기준이 명확하고, **상태 변경에 따른 UI 반응**이 잘 연결되어 있는가?

ohgus committed:
refactor: errorMessage type 수정

- validation 공통 type 추출
- validation 에러 메시지 null 반환 추가

ohgus committed:
docs: step2 기능 요구사항 정리

- 동적 UI에 관한 내용은 추후 업데이트

ohgus committed:
feat: password 입력 영역 구현

ohgus committed:
chore: 오타 수정

ohgus committed:
refactor: AppLayout 기본 높이 제거

ohgus committed:
refactor: CardInputLayout 컴포넌트 수정

- label 옵셔널로 변경
- description margin 추가

ohgus committed:
feat: CustomImage 컴포넌트 구현

ohgus committed:
feat: Select 컴포넌트 구현

ohgus committed:
feat: cardBrand별 카드 색상 적용 로직 구현

ohgus committed:
feat: useBrandSelectInput 훅 구현

ohgus committed:
feat: CardBrandSelectForm 구현

ohgus committed:
chore: 오타수정

ohgus committed:
style: appLayout 최소 높이 지정

ohgus committed:
style: input focus 상황 테두리 색상 조정

ohgus committed:
refactor: change event validation 추가

ohgus committed:
feat: 모든 입력이 올바른지 확인하는 변수 추가

ohgus committed:
feat: Button 컴포넌트 구현

ohgus committed:
feat: Button 스토리북 구현

ohgus committed:
feat: 완료 아이콘 구현

ohgus committed:
chore: react-router 설치

ohgus committed:
feat: 카드 등록 버튼 구현

ohgus committed:
chore: react-router 초기 세팅

ohgus committed:
feat: home, result 페이지 구현

ohgus committed:
refactor: app -> Register로 이동

ohgus committed:
refactor: step 상태 훅으로 추출

ohgus committed:
refactor: errorMessage 관련 로직 수정

모든 값이 유효할 때 에러 미시지를 null로 초기화한다.

ohgus committed:
refactor: 유효 체크 분리

- 모든 값이 유효한지
- 모든 값이 제대로 채워져 있는지

ohgus committed:
feat: 자동 포커스 구현

ohgus committed:
refactor: autoComplete off 추가

ohgus committed:
refactor: flex 0으로 변경

ohgus committed:
feat: cardInput들을 감싸는 CardFormLayout 컴포넌트 추가

ohgus committed:
chore: base url 설정

ohgus committed:
refactor: 년도 숫자 타입으로 변경

ohgus committed:
refactor: 리턴문 간소화

ohgus committed:
refactor: 날짜 입력 값이 채워져도 유효한 날짜가 아니면 다음 스텝으로 넘어갈 수 없게 수정

ohgus committed:
chore: font 경로 수정

ohgus committed:
docs: readme 수정

woowa-euijinkk commented:
오늘 6시 이전에 남기겠습니다!

woowa-euijinkk commented:
버그1)

한번 에러 뜨고 나면, 입력해도 자동으로 input 이 생기지 않네요. blur 해야만 생기군요

<img width="596" alt="image" src="https://github.com/user-attachments/assets/3b707b0e-0695-4479-9099-2cfcc389e677" />

woowa-euijinkk commented on src/components/common/AppLayout/AppLayout.styled.tsx:
```diff
@@ -6,7 +6,7 @@ export const StyledAppLayoutContainer = styled.div`
   width: 100%;
   max-width: 600px;
-  height: 100dvh;
+  min-height: 100dvh;
```
dvh vs vh 어떻게 다른가요?~

ohgus replied:
둘의 차이를 간략하게만 알고 있어 이번 기회에 다시 공부하며 정리해보았습니다 😅

vh와 dvh는 모두 viewport 단위를 기준으로 높이를 계산한다는 공통점이 있지만 기준이 되는 viewport의 정의가 다르며, 이로 인해 레이아웃 안정성에 큰 차이가 발생합니다.

### vh

- 브라우저 전체 viewport 높이를 기준으로 동작한다. (주소창, 탭 바 등 포함)
- 실제로 렌더링되는 시점에 주소창이 열려 있기 때문에 실제로 보이는 영역은 100vh보다 작다.
- 사용자가 스크롤을 내려 주소창이 사라지면, 브라우저가 vh 값을 다시 계산하고 레이아웃이 다시 렌더링된다.
- 다시 렌더링되는 과정에서 **요소가 갑자기 밀리거나 늘어나는 UI 튐 현상**이 발생할 수 있다.

### dvh

- “**현재 보이는 화면 높이**”, 즉 주소창을 제외한 순수 가시 영역을 기준으로 계산한다.
- 사용자가 어떤 상태에서 페이지를 보고 있든, 정확한 높이로 UI를 구성할 수 있다.
- 주소창이 사라지는 경우에도 브라우저는 dvh를 현재 보이는 영역으로 자연스럽게 업데이트하며, 별도의 강제 리렌더링 없이 레이아웃을 부드럽게 확장한다.

데스크탑만을 타겟으로 하는 서비스라면 dvh의 도입 필요성은 낮다고 생각합니다.
그러나 모바일 환경까지 고려해야 하는 경우, dvh는 뷰포트 변화에 따른 예측 불가능한 레이아웃 튐 현상을 사전에 방지할 수 있다는 점에서,
UI 안정성과 사용자 경험 측면에서 확실한 이점을 제공할 수 있다고 생각합니다 👍

woowa-euijinkk commented on src/components/common/Image/CustomImage.styled.tsx:
```diff
@@ -0,0 +1,8 @@
+export const StyledCustomImage = styled.img<Props>`
+  width: ${({ width }) => width || '100%'};
+  height: ${({ height }) => height || '100%'};
+`;
```
이름이 모호합니다. custom 이라고 하여, 들어갔는데 width, height 밖에 존재하지 않네요. 더 확장할 생각을 하고 구현한 것인가요?

width, height 를 css props 로 넣어주기 귀찮아서 이렇게 한 것이라면, 다른 여러 방안이 가능하겠습니다!

ohgus replied:
제가 커스텀 이미지 컴포넌트를 사용한 이유는 css props를 넘겨주기 귀찮아서 한 것은 아닙니다 😅

처음 이 컴포넌트를 만들게 된 이유는, 기존 코드에서 HTML 요소를 직접 사용하는 경우가 거의 없었기 때문에
UI 구성 요소에 일관된 추상화를 적용하고자 하는 목적이었습니다.
그래서 단순한 img 태그라도 별도의 컴포넌트로 분리하여 사용하려고 했습니다.

하지만 시지프의 말씀처럼, 해당 컴포넌트는 역할에 비해 너무 일반적이며, 실제로 캡슐화된 로직이나 의미도 부족한 구조였습니다.
결과적으로 컴포넌트로 분리한 명확한 이유가 존재하지 않는 상태였고, 이로 인해 오히려 코드의 의도를 흐리는 추상화가 되었다고 판단했습니다.

그래서 해당 컴포넌트는 제거하고, 사용하는 곳에서는 목적에 맞게 img 태그를 직접 사용하는 방식으로 변경했습니다.
불필요한 추상화를 줄이고, 의도가 분명한 컴포넌트만 유지되도록 구조를 단순화했습니다!

60bca654e7eccdf961170988bb3ea01dc26f0795

woowa-euijinkk commented on src/components/common/Select/index.tsx:
```diff
@@ -0,0 +1,41 @@
+  onToggle: () => void;
+};
+
+export const Select = ({
+  options,
+  placeholder,
+  isOpen,
+  selectedItem,
+  onItemSelect,
+  onToggle,
+}: SelectProps) => {
```
isOpen, onToggle 을 props 로 받는 이유가 무엇인가요? 

- html 의 select 인터페이스 참고해보아요. 
- 디자인시스템 라이브러리의 Select 를 참고해보아요 (https://www.radix-ui.com/primitives/docs/components/select, https://www.chakra-ui.com/docs/components/select)

ohgus replied:
isOpen, onToggle 같은 경우 select 내부에서 직접 관리하는 방향의 수정을 말씀하신게 맞을까요?

ohgus replied:
수정했습니다! 
혹시 다른 방향을 말씀하신거면 알려주세요 ㅎ

03499802797ce6a599e38319364ce14ed0d3cf4b

woowa-euijinkk replied:
네! 내부로 들어가니까 훨씬 괜찮지 않나요?!

woowa-euijinkk commented on src/components/common/Select/index.tsx:
```diff
@@ -0,0 +1,41 @@
+      )}
+    </StyledSelectContainer>
+  );
+};
```
컴포넌트 설계가 나쁘진 않습니다.

다만, 오거스의 코드와, 위에 첨부한 radix-ui, charka 가 제시한 인터페이스와의 차이점이 무엇인가, 이해하는 것은 중요하겠습니다. 

정답은 없습니다!

ohgus replied:
시지프가 보내주신 라이브러리를 보고 제가 생각한 차이점을 정리해봤습니다.

### 조합형 구조 (Composable Components)
컴포넌트 간 의존성 없이 역할 단위로 분리되어, 개발자가 필요한 구성 요소만 조립해서 사용할 수 있도록 설계되어 있다.

### 명시적인 구조와 유연한 API
컴포넌트 트리만 봐도 구조가 드러나며, 커스터마이징 범위가 넓고 다양한 요구사항에 대응할 수 있는 확장성을 제공한다.

### Context 기반의 암묵적 상태 공유
공통된 상태는 상위 컴포넌트(Context)를 통해 내부적으로 암묵적으로 공유되며, 하위 컴포넌트는 별도의 props 전달 없이 상태에 접근할 수 있다.
이를 통해 외부에서 상태를 일일이 주입하지 않아도 되고, 구성 요소 간의 느슨한 결합과 높은 재사용이 가능하다.

반면, 제가 구현한 컴포넌트는 부모가 모든 상태를 갖고 자식에게 내려주는 방식이어서 상태 흐름이 더 명확한 대신, 구조적으로 유연함은 다소 부족한 형태라고 생각합니다.

woowa-euijinkk commented on src/components/features/Buttons/SubmitButton.tsx:
```diff
@@ -0,0 +1,30 @@
+      left="0"
+      right="0"
+      bottom="0"
+      onClick={() => navigate('/result', { state: { cardType, cardNumber } })}
```
SubmitButton 의 역할이 모호하네요.

이름이 SubmitButton 이고, buttonText 는 props 로 받아서, 마치 재사용 가능한 것처럼 보이지만,
onClick 에서 naviagte('/result') 로 쓰고 있어, 결국 한 곳에서밖에 사용하지 못하는 컴포넌트가 됩니다.

명확한 결정을 하면 좋습니다. 재사용이 필요한 컴포넌트 인가? 아닌가? 
아니라면 과감하게 로직을 내부로 결합시키는 것이 중요하겠습니다!
고민해보아요!

woowa-euijinkk commented on src/routes/index.tsx:
```diff
@@ -0,0 +1,29 @@
+import { Register } from './pages/Register';
+import { Result } from './pages/Result';
+
+const router = createBrowserRouter(
+  [
+    {
+      path: '/',
+      element: <Home />,
+    },
+    {
+      path: '/register',
+      element: <Register />,
+    },
+    {
+      path: '/result',
+      element: <Result />,
+    },
+  ],
+  {
+    basename: '/react-payments/',
+  }
+);
+
+export const Router = () => {
+  return <RouterProvider router={router} />;
+};
```
👍
아래처럼 사용할수도 있답니다. react-router 에서 둘다 지원되는 인터페이스로 보이네요. 어떤게 더 나을지 고민해보아요.

```tsx
export function PageRouter() {
  return (
    <Routes>
      <Route index element={<CartPage />} />
      <Route path={ROUTE_PATH.PAYMENT_CHECK} element={<PaymentCheckPage />} />
    </Routes>
  );
}

export const ROUTE_PATH = {
  PAYMENT_CHECK: '/payment/check',
};

```

ohgus replied:
시지프가 제안해주신 방법의 대해서는 몰랐습니다 😅
과거 팀플의 팀원이 이렇게 사용하는 것을 본 기억이 있어서 그 방식으로 구현했는데 다른 방향이 있었다면 둘 중에 더 좋은게 무엇인지 고민해봤을 것 같습니다.

아래는 제가 찾아본 둘의 차이입니다.

### createBrowserRouter 방식
객체 기반의 선언형 라우팅을 제공하며, RouterProvider를 통해 라우팅을 주입한다.

최신 React Router 기능들(예: loader, action, errorElement)과의 통합성이 좋고, 타입스크립트 기반에서 타입 안전성과 라우트 구조 추론이 뛰어나다.
다만 JSX가 아닌 객체 배열 기반이라 시각적 구조 이해가 어렵고, 간단한 라우팅 구성에는 오히려 과할 수 있다.

**사용하면 좋은 경우**
대규모 앱, 데이터 로딩 최적화가 필요할 때, 타입 안정성을 중시할 때

### Routes/Route 컴포넌트 방식
JSX를 그대로 사용하기 때문에 직관적으로 트리 구조를 파악할 수 있고,
중첩 라우팅 역시 코드만 보고 시각적으로 이해하기 용이하다.
하지만 최신 데이터 로딩 API 등을 활용하려면, 해당 기능을 별도로 구성해야 하기 때문에 기능 통합성에서는 createBrowserRouter보다 약한 구조이다.

**사용하면 좋은 경우**
간단한 프로젝트, 작관적인 코드를 선호할 때

결과적으로 이번 미션의 라우팅 구조는 비교적 단순하고 데이터 로딩 처리 등 복잡한 요구사항이 없기 때문에, 두 방식 중 어떤 방식을 사용했더라도 기능적 차이가 크지 않았을 것 같습니다. 이런 경우에는 구조적 선호도에 따라 선택하는 것도 괜찮을까요?
시지프는 어떤 방식을 더 선호하시나요?

woowa-euijinkk replied:
오오 잘 정리해주셨군요.
사실 저는 오거스가 하신 방식이 있는줄 몰랐습니다 ㅋㅋ 저 방식도 괜찮겠다 싶네요.

저도 기능적 차이가 크진 않아 보여서 선호도에 따라 다를 것 같네요!

woowa-euijinkk commented on src/routes/pages/Home.tsx:
```diff
@@ -0,0 +1,28 @@
+import { Flex } from '@/components/common/Flex';
+import { Text } from '@/components/common/Text';
+
+export const Home = () => {
+  const navigate = useNavigate();
+
+  return (
+    <AppLayout>
+      <Flex direction="column" gap="20px" width="80%" margin="0 auto">
+        <Text variant="Title" fontWeight="bold">
+          우테코 자동 결제 서비스 이용을 위해
+        </Text>
+        <Text variant="Title" fontWeight="bold">
+          카드 등록 페이지로 이동해주세요.
+        </Text>
+        <Button height="60px" borderType="rounded" onClick={() => navigate('/register')}>
+          <Text variant="Title" fontWeight="medium" color="white">
+            카드 등록하러 가기
+          </Text>
+        </Button>
+      </Flex>
+    </AppLayout>
+  );
+};
```
👍

woowa-euijinkk commented on src/routes/pages/Register.tsx:
```diff
@@ -0,0 +1,130 @@
+    canMoveNextFromPassword,
+  });
+
+  const isAllClearInput =
```
변수명이 모호하네요.
AllClear? 다 지운다는 의미인가? 라고 생각이 들었답니다. (계산기에서 AllClear 는 다 지운다는 의미이니..)

ohgus replied:
변수화 수정 단계에서 isUserCanSubmit으로 수정했습니다!

woowa-euijinkk replied:
canSubmit 으로 간단하게 써도 좋겠네요 ㅎㅎ

변수명에 is/can/should 등이 prefix 로 들어갈 수 있으니까요

woowa-euijinkk commented on src/hooks/useCardInput.ts:
```diff
@@ -56,9 +81,35 @@ export const useCardInput = (type: CardFormFiledType) => {
+  return value.every((item) => item.value.length === CardInputTypeOptions[type].valueLength);
+}
+
+function canMoveNextInput(value: CardInputType[], type: CardFormFiledType) {
```
`can~` 는 변수에 쓰이는 느낌이긴 하네요.
함수는 `checkCanMove~` 같이 동사가 붙으면 좋지 않을까 싶습니다!

ohgus replied:
83743b9fb86482a9e894604c8102f8e49d4ed90d

수정했습니다!

woowa-euijinkk commented on src/hooks/useCardInput.ts:
```diff
@@ -56,9 +81,35 @@ export const useCardInput = (type: CardFormFiledType) => {
+  };
 };
+
+function isAllValidInput(value: CardInputType[]) {
+  return value.every((item) => item.isValid);
+}
+
+function isAllFilledInput(value: CardInputType[], type: CardFormFiledType) {
+  return value.every((item) => item.value.length === CardInputTypeOptions[type].valueLength);
+}
```
이 함수를 외부에 둔 이유가 궁금합니다! 
useCardInput 의 내부에 두는 것도 방법인 것 같아서요!

ohgus replied:
사실 당시에 조금 급하게 구현했는데 다른 코드들과 같이 있으면 제가 놓치는 부분이 있을 것 같아서 외부에 구현하고 다시 안으로 넣어야지 했던 것을 그대로 커밋해버렸습니다... 🥲

수정했습니다!
de87b5aab168471b06d69d4a76c0239c8f5911f6

woowa-euijinkk commented on src/hooks/useCardInput.ts:
```diff
@@ -28,11 +33,32 @@ export const useCardInput = (type: CardFormFiledType) => {
+    const inputValue = e.target.value;
+    const { isValid: isValidChange, errorMessage } = validateInputChange(inputValue);
+
+    if (!isValidChange) {
+      setErrorMessage(errorMessage);
+
+      setValue((prev) => {
+        const newArr = [...prev];
+        newArr[index].isValid = isValidChange;
+        return newArr;
+      });
+
+      return;
+    }
+
```
valid 하지 않을땐, value 를 update 해주지 않고 있는데요. 문제가 없을까요?

ohgus replied:
유효하지 않은 값이 들어왔을 때 value를 업데이트하지 않은 이유는,
입력값이 UI에 반영되지 않도록 설계함으로써 사용자가 잘못된 값을 시도했을 때 즉시 차단되도록 하기 위함이었습니다.
즉, 유효하지 않은 값은 애초에 화면에 표시되지 않도록 하여 일관된 상태를 유지하는 방식을 선택했습니다.

그리고 시지프가 말씀하신 “문제가 없을까요?”라는 피드백에 대해
써밋과 대화를 나누며 추론해 본 결과,
혹시 실제 값은 업데이트되지 않았지만 에러 메시지는 출력되는 상황이
사용자 입장에서 혼란스럽게 느껴질 수 있다는 UX 측면의 문제를 말씀하신 게 아닐까 생각해보게 되었습니다.
(혹시 제가 다르게 이해한 부분이 있다면 말씀해주시면 감사하겠습니다!)

woowa-euijinkk replied:
네 맞습니다!
UX 관점에서 여러 방향으로 생각해볼 수 있어서 말씀드렸습니다
👍

woowa-euijinkk commented on src/routes/pages/Register.tsx:
```diff
@@ -0,0 +1,130 @@
+    canMoveNextFromCVC &&
+    canMoveNextFromExpireDate &&
+    canMoveNextFromCardNumber &&
+    selectedBrand !== null;
```
이것도 변수화하면 어떨까요?

ohgus replied:
```ts
const isAllValidInput =
    canMoveNextFromPassword &&
    canMoveNextFromCVC &&
    canMoveNextFromExpireDate &&
    canMoveNextFromCardNumber &&
    selectedBrand !== null;

  const isUserCanSubmit = step >= 4 && isAllValidInput;
```

시지프가 제안하신 부분은, 유효성 검증 로직을 의미 단위로 분리해
조건식의 가독성과 유지보수성을 높이자는 의도로 이해했습니다.
(제가 잘못 이해했다면 알려주세요!)

이에 따라 각 필드의 유효성 검사를 isAllValidInput이라는 변수로 추출하고,
현재 step 조건과 조합하여 최종적으로 유저가 버튼을 누를 수 있는 상태를 isUserCanSubmit으로 선언했습니다.

이렇게 나누면 조건들의 의미가 명확해지고, 이후 조건이 늘어나거나 바뀔 때에도 수정 범위를 제한할 수 있어
코드의 의도를 명확히 전달하는 데 도움이 될 것 같습니다!

ohgus replied:
bbe6b0a717ae89cc95951ec43c110524ee6eec50

woowa-euijinkk commented on src/hooks/useFormStep.ts:
```diff
@@ -0,0 +1,49 @@
+}: UseFormStepProps) => {
+  const [step, setStep] = useState(0);
+
+  useEffect(() => {
+    if (step === 3 && canMoveNextFromCVC) {
+      setStep(step + 1);
+      return;
+    }
+
+    if (step === 2 && canMoveNextFromExpireDate) {
+      setStep(step + 1);
+      return;
+    }
+
+    if (step === 1 && selectedBrand) {
+      setStep(step + 1);
+      return;
+    }
+
+    if (step === 0 && canMoveNextFromCardNumber) {
+      setStep(step + 1);
+      return;
+    }
```
ㅋㅋ ㅜㅜㅜ 
하나씩 하드코딩 해버렸군요. 더 좋은 아이디어를 구상해봅시다! 
레퍼런스 참고하고, 구상하는 과정에서 많이 배우실 수 있을거에요. 꼭 추천드립니다

woowa-euijinkk replied:
지금의 문제점은 아시겠죠?

순서가 바뀐다면, 변경 포인트가 많습니다. 버그 확률이 매우 높아요.

ohgus replied:
혹시 시지프가 말씀하신 레퍼런스가 무엇인지 알려주실 수 있을까요?
제가 수정한 방식이 좋은 방식인지 확신이 들지 않는 것 같습니다 ㅜ

410140710cab916ecac5fabf4be266beb5f615c4

woowa-euijinkk commented on src/routes/pages/Register.tsx:
```diff
@@ -0,0 +1,130 @@
+            onItemSelect={handleBrandSelect}
+            onToggle={handleToggleSelect}
+          />
+        )}
```
step 코드 구조가 좋진 않지만, 일관되게 짜두셔서, 수정이 용이할 것 같네요!👍

ohgus replied:
c24dc93a54dcbd7c019e699c9dbc329709a68853

수정했습니다!

woowa-euijinkk commented on src/routes/pages/Register.tsx:
```diff
@@ -0,0 +1,130 @@
+      {isAllClearInput && (
+        <SubmitButton
+          cardType={selectedBrand}
+          cardNumber={cardNumbers[0].value}
```
props 가 조금 이상합니다 ㅎㅎ
onClick 을 받는게 더 자연스럽지 않을까 생각이 드네요. SubmitButton 의 역할이 무엇인지 생각해보아요.
그 역할을 수행함에 있어서 꼭 cardType, cardNumber를 직접 알아야할까?! 고민해봅시다

ohgus replied:
텍스트 children과 같이 수정했습니다!

df7c77ada2674a57837a7859c42dabb2e03f7a5a

woowa-euijinkk commented on src/routes/pages/Register.tsx:
```diff
@@ -0,0 +1,130 @@
+        <SubmitButton
+          cardType={selectedBrand}
+          cardNumber={cardNumbers[0].value}
+          buttonText="확인"
```
children 으로 하면 어떨까요?
우리가 자연스러운 interface 를 만드는게 중요합니다
`<button>확인</button>` 이라고 html 을 쓰는 것과 같이요.

woowa-euijinkk commented on src/routes/pages/Result.tsx:
```diff
@@ -0,0 +1,33 @@
+  const location = useLocation();
+  const navigate = useNavigate();
+
+  const { cardType, cardNumber } = location.state;
```
history state 를 넘기셨군요.
이건 타입도 없고, 이전 페이지에서 이런 값이 넘어올 것이야! 하고 믿을 수밖에 없는 코드인데요.

어떻게 하면 안정성을 높일 수 있을까요?

ohgus replied:
<img width="530" alt="image" src="https://github.com/user-attachments/assets/3090e931-8312-4296-9bb7-602849167d1d" />

### 1. 명시적 타입 정의
ResultProps 타입을 선언하고 location.state에 타입 단언을 통해 적용했습니다.
이를 통해 코드상에서 cardType, cardNumber가 존재한다는 의도를 표현했습니다.

### 2. fallback UI
혹시라도 값이 존재하지 않을 경우를 대비해 조건 분기 렌더링을 추가했습니다.

ohgus replied:
adb07b7e5a9bd5ca755484702bc72f6925be1ffb

woowa-euijinkk replied:
이부분은 TS 측면에서 고민할 거리가 참 많은데요.

문제점과 여러가지 해결책을 말씀드릴게요. 학습할 숙제를 드리겠습니당!

## 문제점
### 1. 타입단언이 부적절함.

`const { cardType, cardNumber } = location.state as ResultProps;`
ResultProps 로 온다는 보장이 없습니다. 타입 단언을 함부로 사용하면 안 됩니다. 오거스는 이 코드를 이해하고 있기 때문에, ResultProps 로 오면 잘 처리하고, 그렇지 않으면 아래에서 분기를 쳐야지. 라고 판단하고 코드를 짜신거죠. 다른 협업자가 보면 맥락을 이해하기 어렵고 버그 가능성이 생깁니다.

즉, 단언은 최소한으로 하는게 좋으며, 단언을 하더라도 제대로 단언해야 합니다.

```ts
type ResultProps = {
  cardType?: CardBrandsType;
  cardNumber?: string;
};

const cardInfoFromState = location.state as ResultProps | undefined;

// ...

if (cardInfoFromState == null) {
  return <에러처리 />
}

if (cardInfoFromState.cardType == null || cardInfoFromState.cardNumber == null) {
  return <에러처리 />
}

// 정상 케이스 작성
```

### 2. 타입단언 쓰지 않기
타입 단언을 쓰지 않고도 해결할 수 있습니다. location.state 는 사실상 unknown 타입입니다. 알려지지 않은거죠. 어떻게 들어올지 전혀 예측할 수 없습니다. 이때 type guard 을 사용하면 좋고, 그 중 하나로 `User-Defined Type Guards` 를 사용하시면 좋습니다.

타입단언을 쓰지 않고도, TypeScript, JavaScript 시스템 내에서 에러를 잡아낼 수 있습니다. 타입단언은 TypeScript 시스템을 무시하고, 내 마음대로 선언하겠다는 의미이죠. 가능한 한 TypeScript 시스템 내에서 처리해주는 것이 좋겠죠.

`User-Defined Type Guards` 를 학습해보아요. 종종 쓰입니다.

spoyodevelop replied:
satisfies를 써도 괜찮게 동작할까요? 궁금하네요...

woowa-euijinkk replied:
좋은 의문이네요 마빈.

satisfied 는 컴파일 타임에서 타입을 체크하죠.

지금 상황은 외부(history)에서 state 가 넘어오는 예시이기 때문에, unknown 타입이 오는거고, 정확히는 "런타임 type guard" 가 필요합니다.

위 케이스에서 as 를 satisfied 로 바꾼다고 하여, 바뀌는 건 없을 것 같네요.

spoyodevelop replied:
window.history에서 넘겨 오게 구현부가 짜여져 있을까요... 그렇다면 unknown 타입이 온다고 생각을 하는게 좋겠고, 사실상 외부에서 데이터가 넘어 오는 경우에는 unknown을 쓰는 것이 맞겠네요.
const data = await response.json() as unknown; 같이 fetch로 가져 오는것도 unknown으로 하고 런타임 type guard가 필요할까요?

woowa-euijinkk replied:
좋은 질문이네요.

API 도 런타임 체크를 해야하지 않겠느냐? 서버에서 원하는 데이터가 온다고 100% 확신할 수 있느냐? 라는 의문도 있는데요.

그래서 들어보셨을지 모르겠는데, [zod](https://github.com/colinhacks/zod) 라는 라이브러리가 아주 많이 쓰입니다.
많은 회사에서 사용하고 있어요.

woowa-euijinkk commented on src/components/features/CardFormFiled/CardPasswordForm.tsx:
```diff
@@ -0,0 +1,56 @@
+            ref={refs[index]}
+            autoComplete="off"
+          />
+        ))}
```
얘는 왜 반복문을 돌고있나용?

ohgus replied:
541b53e13ebac832b382cac7db989fe3b5ce00b7

ref도 같이 수정했습니다!

현재 password나 cvc는 단일 input만 렌더링되지만,
다른 form 필드들과의 코드 구조의 일관성을 유지하기 위해 map을 사용하는 방식을 선택했습니다.

다만, 말씀하신 것처럼 해당 필드가 항상 단일 input으로 고정되는 것이 명확하다면,
단순한 하드코딩 방식으로 렌더링하는 것이 더 직관적이고 불필요한 반복 처리를 피할 수 있지 않을까 하는 생각도 듭니다.

시지프라면 이런 경우에는 하드 코딩으로 구현하는 방식도 괜찮다고 생각하시나요?

woowa-euijinkk replied:
아하. 그런 의도였군요.

새로운 협업자가 cvc 를 수정하러 올 때, 반복문이 있는 것을 보면 혼란스럽지 않을까 싶습니다. 모든 코드 베이스를 이해하고 있다면야, 의도를 파악할 수 있겠지만요.

이 부분은 저는 과도한 일관성이라고 생각이 드네요요. CVC 나 password 가 여러개 input 을 가질 가능성이 없으니까요. 만약 그런 일관성을 챙기고 싶다면, 반복문 도는 함수를 분리해주는 것이 적절하겠네요. <Inputs /> 같은 컴포넌트로 만들어서요. 시스템화를 해보아요.

woowa-euijinkk commented on src/components/features/CardFormFiled/CVCForm.tsx:
```diff
@@ -7,12 +7,15 @@ import { Flex } from '@/components/common/Flex';
 } & CardFormFiledProps;
 
 export const CVCForm = ({ cvcNumber, errorMessage, onCardInputChange, onCardInputBlur }: Props) => {
+  const { refs } = useFocus(cvcNumber.length);
```
얘는 그냥 number(ex. 343) 가 들어오는데, number.length 만큼의 ref 를 만들 필요가 있나요?

woowa-euijinkk commented on src/hooks/useFocus.ts:
```diff
@@ -0,0 +1,15 @@
+    if (refs[0].current) {
+      refs[0].current.focus();
+    }
+  }, []);
```
함수화를 똑같이 신경써보아요. 코드를 이해하는데 시간이 조금 소요되었어요 ㅎㅎ

이렇게 네이밍 붙여서, useEffect 를 쓸 수도 있어요. 

```suggestion
  useEffect(function focusOnFirstRef() {
        if (refs[0].current) {
            refs[0].current.focus();
        }
    }, []);
```

ohgus replied:
수정했습니다!

36bd348eaea9dabd1d76e0e9f968eb4f7a3fea0d

woowa-euijinkk commented on src/validations/validateExpireDate.ts:
```diff
@@ -1,16 +1,10 @@
+export const validateExpireDate = (value: string, key: ExpireDateInputKey): ValidateReturnType => {
   const expireDateRegex = new RegExp(`^\\d{2}$`);
-  const currentYear = new Date().getFullYear().toString().slice(-2);
+  const currentYear = new Date().getFullYear() % 100;
```
왜 로직을 바꾸셨나요?~

ohgus replied:
해당 변수를 사용하는 목적을 다시 확인해보니, `currentYear`는 문자열 비교가 아닌 숫자 간의 크기 비교에 사용되고 있었습니다.
이처럼 값의 목적이 수치적 비교에 있다면, 이를 문자열로 선언하는 것은 타입과 의미 간 불일치라고 판단했습니다.

따라서 변수의 의도와 사용 방식이 일관되도록 숫자형으로 선언하는 것이 더 자연스럽고,
불필요한 문자열 변환 비용도 줄일 수 있어 해당 로직을 수정하게 되었습니다.

woowa-euijinkk commented on src/validations/validateExpireDate.ts:
```diff
@@ -23,12 +17,12 @@ export const validateExpireDate = (
   }
 
-  if (key === 'year' && parseInt(value) < parseInt(currentYear)) {
+  if (key === 'year' && parseInt(value) < currentYear) {
```
변수화해볼 수 있을거같아요. 더 읽기쉽게!

```suggestion
  if (key === 'year') {
    const isPastYear = parseInt(value) < currentYear;
    if (isPastYear) {

    }

```

ohgus replied:
수정했습니다!

fa93264b35bf1665db8d13835b3fb784a3e9547a

woowa-euijinkk commented on src/stories/CardPreview.stories.tsx:
```diff
@@ -31,6 +31,7 @@ export const Default: Story = {
       month: { value: '12', isValid: true },
       year: { value: '34', isValid: true },
     },
+    cardBrand: '현대카드',
```
현대카드가 사라진다면 어떻게 되나용?

ohgus replied:
초기 값은 null로 초기화하고 상수로 정의된 카드 브랜드를 options로 제공했습니다!

558155adcd93af980d1f52eb4d777e3dd31e0163

woowa-euijinkk commented on src/components/features/CardPreview/CardPreview.styled.tsx:
```diff
@@ -2,9 +2,31 @@ import styled from '@emotion/styled';
+  | '우리카드'
+  | '롯데카드'
+  | '하나카드'
+  | '국민카드';
```
이 코드와 중복되는데 어떻게 하면 좋을까요?

```ts
const cardBrandOptions = [
  'BC카드',
  '신한카드',
  '카카오뱅크',
  '현대카드',
  '우리카드',
  '롯데카드',
  '하나카드',
  '국민카드',
];
```

ohgus replied:
feb9c8128fe04efef32204a27b68d7d62b9f3822

이미 정의되어 있는 `cardBrandColors`에서 `key`를 추출하여 `CardBrandType` 과 `cardBrandOptions`에서 재사용할 수 있게 수정하였습니다.

woowa-euijinkk replied:
keyof 를 잘 적용해주셨군요 ㅎㅎ !

woowa-euijinkk reviewed (changes requested):
오거스 리뷰가 늦어 죄송해요.
2단계 잘해주셨군요!
전반적으로 잘 해주셨고, 좋은 학습 과정에 있다고 생각합니다. 조금만 다듬으면 더 좋은 코드가 될거같아요 ㅎㅎ 
리뷰 남긴 포인트 위주로 생각해보면 성장에 많이 도움이 되실거에요!

ohgus committed:
refactor: isOpen, onToggle을 select 컴포넌트 내부로 이동

ohgus committed:
refactor: callback 함수를 기명 함수로 변경

ohgus committed:
refactor: validateExpireDate 조건문을 변수화해 가독성 향상

ohgus committed:
refactor: 중복된 cardBrand 리팩토링

ohgus committed:
test: 카드사 선택의 초기값을 Null로 변경

ohgus committed:
refactor: cardBrandType을 별도의 파일로 추출

ohgus committed:
refactor: 불필요한 map 제거 & 단일 ref 고정

ohgus committed:
refactor: type 단언 추가

ohgus committed:
refactor: result page 타입 단언 추가 & fallback UI 추가

ohgus committed:
refactor: SubmitButton 수정

- 텍스트를 children으로 받을 수 있게 수정
- submit에 대한 동작을 내부에서 정의하지 않고 onSubmit으로 받을 수 있게 변경

ohgus committed:
refactor: 조건문을 의미가 비슷한 그룹끼리 묶어서 분리

ohgus committed:
refactor: 함수 이름을 동사형으로 변경

ohgus committed:
refactor: 역할이 모호한 CustomImage 컴포넌트 제거

- 각 역할에 맞는 컴포넌트 제작
- SelectArrow
- CompleteIcon

ohgus committed:
refactor: ROUTE_PATH 정의

ohgus committed:
refactor: hook에서만 사용하는 함수를 훅 내부로 이동

ohgus committed:
test: 스토리에 컴포넌트 변경사항 반영

ohgus committed:
refactor: form 별 스텝을 상수로 추출

ohgus committed:
refactor: FORM_STEP_NUMBERS로 이름 변경

- password 추가

ohgus committed:
refactor: step별 컴포넌트 렌더링 방식 수정

ohgus committed:
refactor: step 업데이트 로직 간소화

ohgus committed:
refactor: password input value를 설정해 state와 동일한 값을 유지할 수 있게 수정

ohgus committed:
fix: 에러 상태에서 올바른 값을 입력해도 에러가 지워지지 않는 문제 수정

에러가 지워지지 않아서 다음 스텝으로 넘어가지 않는 문제가 발생했는데 이를 수정했습니다.

ohgus committed:
refactor: 날짜 관련 예외처리를 체인지 이벤트로 위임

블러 이벤트는 숫자, 길이에 대한 체크만 진행

ohgus committed:
refactor: 핸들러 처리 방식 수정

블러 이벤트는 value 상태를 업데이트하지 않는다.
체인지 이벤트에서 전반적인 상태를 관리한다.

ohgus commented:
안녕하세요 시지프!
리뷰 반영이 늦었습니다 🥲

공부를 병행하면서 리뷰를 반영하고 코멘트를 다느라 평소보다 시간이 더 걸린 것 같습니다.
좀 어려운 부분들도 많았지만 수정을 거치면서 정말 재밌었고 많이 배웠다고 생각합니다 😁
이번에도 꼼꼼한 리뷰를 남겨주셔서 감사합니다 🙇

그리고 제대로 이해되지 않은 부분은 코멘트로 다시 남겼습니다!

woowa-euijinkk reviewed (approved):
많은 코드를 수정해주셨군요!

끝까지 해주시느라, 오거스 정말 고생많으셨어요. 잘해주셨습니다. 많이 성장하셨으면 좋겠습니다!

코멘트 확인만 해주세요~
특히 이 메시지는 TS 학습에 주요한 부분이니, 꼭 확인 부탁드려요~~
https://github.com/woowacourse/react-payments/pull/495#discussion_r2083795228

머지하도록 하겠습니다. 고생하셨습니다!
